spark/src/main/scala/com/gu/tableversions/spark/VersionedDataset.scala

```diff
@@ -77,42 +86,32 @@ object VersionedDataset {
       _ <- metastore.update(table.name, metastoreUpdate)
 
     } yield (latestTableVersion, metastoreUpdate)
-
-  private def operationsForPartitions(
-      partitions: List[Partition],
-      version: Version): List[TableVersions.TableOperation] = {
-    if (partitions == List(Partition.snapshotPartition))
-      List(AddTableVersion(version))
-    else partitions.map(partition => AddPartitionVersion(partition, version))
   }
 
   /**
     * Get the unique partition values that exist within the given dataset, based on given partition columns.
     */
   private[spark] def partitionValues[T](dataset: Dataset[T], partitionSchema: PartitionSchema)(
       implicit spark: SparkSession): List[Partition] = {
-    if (partitionSchema == PartitionSchema.snapshot) {
-      List(Partition.snapshotPartition)
-    } else {
-      // Query dataset for partitions
-      // NOTE: this implementation has not been optimised yet
-      val partitionColumnsList = partitionSchema.columns.map(_.name)
-      val partitionsDf = dataset.selectExpr(partitionColumnsList: _*).distinct()
-      val partitionRows = partitionsDf.collect().toList
-
-      def rowToPartition(row: Row): Partition = {
-        val partitionColumnValues: List[(Partition.PartitionColumn, String)] =
-          partitionSchema.columns zip row.toSeq.map(_.toString)
-
-        val columnValues: List[Partition.ColumnValue] = partitionColumnValues.map {
-          case (partitionColumn, value) => Partition.ColumnValue(partitionColumn, value)
-        }
-
-        Partition(columnValues)
-      }
+    // Query dataset for partitions
+    // NOTE: this implementation has not been optimised yet
+    val partitionColumnsList = partitionSchema.columns.map(_.name)
+    val partitionsDf = dataset.selectExpr(partitionColumnsList: _*).distinct()
+    val partitionRows = partitionsDf.collect().toList
+
+    def rowToPartition(row: Row): Partition = {
+      val partitionColumnValues: List[(Partition.PartitionColumn, String)] =
+        partitionSchema.columns zip row.toSeq.map(_.toString)
 
-      partitionRows.map(rowToPartition)
+      val columnValues = partitionColumnValues.map(Partition.ColumnValue.tupled)
+
+      columnValues match {
+        case head :: tail => Partition(head, tail: _*)
+        case _            => throw new Exception("Empty list of partitions not valid for partitioned table")
+      }
```

Comment:
This runtime error is needed because `PartitionSchema` is used for both the partitioned and snapshot cases at the moment hence the type system doesn't enforce an empty list of partition columns. One for a later refactoring, we think...